extensions/data-plane/data-plane-google-bigquery/src/main/java/org/eclipse/edc/connector/dataplane/gcp/bigquery/pipeline/BigQueryDataSink.java

```diff
@@ -0,0 +1,337 @@
+import org.eclipse.edc.gcp.bigquery.BigQueryTarget;
+import org.eclipse.edc.gcp.common.GcpException;
+import org.json.JSONArray;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.Phaser;
+import java.util.concurrent.TimeUnit;
+
+import static org.eclipse.edc.connector.dataplane.spi.pipeline.StreamFailure.Reason.GENERAL_ERROR;
+import static org.eclipse.edc.connector.dataplane.spi.pipeline.StreamResult.failure;
+import static org.eclipse.edc.connector.dataplane.spi.pipeline.StreamResult.success;
+
+/**
+ * Writes JSON data in a streaming fashion using the BigQuery Storage API (RPC).
+ */
+public class BigQueryDataSink extends ParallelSink {
+    private BigQueryConfiguration configuration;
+    private BigQueryTarget target;
+    private GoogleCredentials credentials;
+    private final Phaser inflightRequestCount = new Phaser(1);
+    private BigQueryWriteClient writeClient;
+    private JsonStreamWriter streamWriter;
+
+    private BigQueryDataSink() {
+    }
+
+    void testAppendSignal() {
+        inflightRequestCount.arriveAndDeregister();
+    }
+
+    @Override
+    protected StreamResult<Object> transferParts(List<DataSource.Part> parts) {
+        if (parts.isEmpty()) {
+            return StreamResult.success();
+        }
+
+        Exception appendException = null;
+        var errorWhileAppending = false;
+        for (var part : parts) {
+            try (var inputStream = part.openStream()) {
+                var mapper = new ObjectMapper();
+                var jsonParser = mapper.createParser(inputStream);
+                var token = jsonParser.nextToken();
+                while (token != null && token == JsonToken.START_ARRAY) {
+                    var element = mapper.readTree(jsonParser);
+                    var jsonString = element.toString();
+                    var page = new JSONArray(jsonString);
+                    append(page);
+                    token = jsonParser.nextToken();
+                }
+
+                if (token != null && token != JsonToken.START_ARRAY) {
+                    throw new IllegalStateException("BigQuery sink JSON array expected");
+                }
+            } catch (Exception exception) {
+                errorWhileAppending = true;
+                appendException = exception;
+                monitor.severe("BigQuery Sink error while appending: ", appendException);
+                break;
+            }
+
+            if (part instanceof BigQueryPart bigQueryPart) {
+                if (bigQueryPart.getException() != null) {
+                    errorWhileAppending = true;
+                    appendException = bigQueryPart.getException();
+                    monitor.severe("BigQuery Sink error from source: ", appendException);
+                    break;
+                }
+            }
+        }
+
+        inflightRequestCount.arriveAndAwaitAdvance();
+        closeSinkStream();
```

Comment:
the way the code is written now, this could potentially re-open the stream and immediately close it again. it would be better to (silently) close the stream here.

Reply:
Done